File: src/main/java/leetcode/groupanagrams_49/Solution.java
package leetcode.groupanagrams_49;

import java.util.*;

public class Solution {

    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> anagramMap = new HashMap<>();

        for (String str : strs) {
            char[] characters = str.toCharArray();
            Arrays.sort(characters);

            String sorted = String.valueOf(characters);

            List<String> anagrams = anagramMap.getOrDefault(sorted, new ArrayList<>());
            anagrams.add(str);

            anagramMap.put(sorted, anagrams);
        }

        return new ArrayList<>(anagramMap.values());
    }
}

File: src/main/java/leetcode/groupanagrams_49/Solution.kt
package leetcode.groupanagrams_49

fun groupAnagrams(strs: Array<String>): List<List<String>> {
    val results: MutableMap<String, MutableList<String>> = mutableMapOf()

    for (str in strs) {
        val sorted = str.toCharArray().sorted().joinToString("")

        results.getOrPut(sorted) { mutableListOf() }

        results[sorted]!!.add(str)
    }

    return ArrayList<List<String>>(results.values)
}
